### DISCUSSION TITLE
Encrypting files using ECDSA

### DISCUSSION THREAD

**Discussion by AngeloMetal** [Autarkysoft/Denovo#4] [Q&A]
I was trying to implement file encryption using Denovo's ECDSA and I would like to know how can this be possible. Can I encrypt the bytes of a file without encrypting its binaries first? I tried to do it with its binaries but its size is getting 8 times bigger so it's not worth it. What I do need is encrypting the bytes of a file straight away. I haven't tried it yet, I just want to know if it's possible.

Also I'm encountering another issue. Once a file is encrypted it must contain its name too right? Encrypting only the content means that the receiver must know the extension of the file right after he/she decrypts it.

**Comment by Coding-Enthusiast** [comment]
> Can I encrypt the bytes of a file without encrypting its binaries first?

That's essentially the same thing, the encryption methods have to get a stream of bytes to split into smaller fixed size blocks and encrypt it to finally return a result.  

> What I do need is encrypting the bytes of a file straight away. I haven't tried it yet, I just want to know if it's possible. 

It is possible. You have to read the file as a stream and pass it to your encryption method. I'm afraid there is no method in Bitcoin.Net that can help with that; but it is not too hard to implement. Here is a good SO link: https://stackoverflow.com/questions/9237324/encrypting-decrypting-large-files-net  
You can replace the `Rfc2898DeriveBytes` with `Autarkysoft.Bitcoin.Cryptography.KeyDerivationFunctions.Scrypt` for a much stronger KDF.

> Once a file is encrypted it must contain its name too right? Encrypting only the content means that the receiver must know the extension of the file right after he/she decrypts it.

The file type or the content type could be communicated in another way (the file name doesn't matter). For example the content could simply be plain-text or the common JSON format like what some wallets use in which case the name or extension of the encrypted file doesn't matter. For instance you could call the file `Foo.encrypted`. You could also define an ecnryption/decryption contract where the content type is indicated by the first 10 bytes.

**Reply by AngeloMetal** [reply]
> It is possible. You have to read the file as a stream and pass it to your encryption method. I'm afraid there is no method in Bitcoin.Net that can help with that; but it is not too hard to implement. Here is a good SO link: https://stackoverflow.com/questions/9237324/encrypting-decrypting-large-files-net
You can replace the `Rfc2898DeriveBytes` with `Autarkysoft.Bitcoin.Cryptography.KeyDerivationFunctions.Scrypt` for a much stronger KDF.

Unfortunately, I don't fully understand what does that function on stackoverflow do and because my philosophy is to always understand the steps, I was thinking of doing it with another way. The bytes of a file represent a string (once they are converted to string). The problem is the conversion. 

For example, this is the file when I open it with a text editor: https://i.imgur.com/VBvQ8bG.png
And this is when I have converted it to string: https://i.imgur.com/VHHaX0R.png

I convert it with `var convertedString = System.Text.Encoding.UTF8.GetString(fileBytes);` and then write it on a file with `File.WriteAllText(@"C:\Users\blabla\lock3.txt", convertedString);`
I guess that going from bytes to string doesn't mean that we can go from string to bytes as well. Straight forward, without converting the bytes to binary/hex first and then to string. I'm probably wrong about this, but I would also like to read your reply and rethink. There must be a way to convert the bytes to string exactly as they were or even better, a file function that instead of string, takes bytes.

**Reply by Coding-Enthusiast** [reply]
Avoid converting the content to anything. You are adding an extra pointless step (since the string will be converted to bytes before it is encrypted) and can lose data during conversion when the encoding uses a smaller set than your data.  
As I said before, just work with the byte array.   
You can also take a look at AES examples on MSDN (https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.aesmanaged?view=net-5.0)
